web/toolbar-geckoview.js

```diff
@@ -0,0 +1,79 @@
+/* Copyright 2023 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @typedef {Object} ToolbarOptions
+ * @property {HTMLDivElement} mainContainer - Main container.
+ * @property {HTMLDivElement} container - Container for the toolbar.
+ * @property {HTMLButtonElement} download - Button to download the document.
+ */
+
+class Toolbar {
+  #buttons;
+
+  #eventBus;
+
+  #toolbar;
+
+  #mainContainer;
+
+  /**
+   * @param {ToolbarOptions} options
+   * @param {EventBus} eventBus
+   * @param {IL10n} _l10n - Localization service.
+   */
+  constructor(options, eventBus, _l10n) {
+    this.#toolbar = options.container;
+    this.#mainContainer = options.mainContainer;
+    this.#eventBus = eventBus;
+    this.#buttons = [{ element: options.download, eventName: "download" }];
+
+    // Bind the event listeners for click and various other actions.
+    this.#bindListeners(options);
+  }
+
+  setPageNumber(pageNumber, pageLabel) {}
+
+  setPagesCount(pagesCount, hasPageLabels) {}
+
+  setPageScale(pageScaleValue, pageScale) {}
+
+  reset() {}
+
+  #bindListeners(options) {
+    // The buttons within the toolbar.
+    for (const { element, eventName, eventDetails } of this.#buttons) {
+      element.addEventListener("click", evt => {
+        if (eventName !== null) {
+          this.#eventBus.dispatch(eventName, { source: this, ...eventDetails });
+        }
+      });
+    }
+  }
+
+  updateLoadingIndicatorState(loading = false) {}
+
+  show() {
+    this.#toolbar.classList.toggle("show", true);
+    this.#mainContainer.addEventListener("scroll", this.hide.bind(this));
+  }
+
+  hide() {
+    this.#toolbar.classList.toggle("show", false);
+    this.#mainContainer.removeEventListener("scroll", this.hide.bind(this));
+  }
```

Comment:
I don't believe that the `removeEventListener`-call works as intended here, since you need to pass a reference to the *original* event-listener function in order to remove it. Note how we handle that in e.g. [the PresentationMode implementation](https://github.com/mozilla/pdf.js/blob/08c8d310d3852b16aad0a511aec9b01216c0f590/web/pdf_presentation_mode.js#L341-L377).

However, we might be able to simplify things a fair bit here by using `once` instead:
```suggestion
    this.#mainContainer.addEventListener(
      "scroll",
      () => {
        this.#toolbar.classList.toggle("show", false);
      },
      { once: true }
    );
  }
```